Fix shared-volume permission issue(s)

codemedic opened the pull request:
- No "data-paths" will be created in the image, in order to stop docker from mucking up permissions
- gpg-s3sync will now check all paths that it works with to make sure it can proceed effectively; if not it bails with descriptive error
- data-generateor split out into a seperate docker file and as a new service in `docker-compose.test.yml`, in order to emulate a real data generation service
- various improvements to the test script
- Added extra docker-ignores

codemedic committed:
Fix shared-volume permission issue(s)

- No "data-paths" will be created in the image, in order to stop docker from mucking up permissions
- gpg-s3sync will now check all paths that it works with to make sure it can proceed effectively; if not it bails with descriptive error
- data-generateor split out into a seperate docker file and as a new service in `docker-compose.test.yml`, in order to emulate a real data generation service
- various improvements to the test script
- Added extra docker-ignores

abulford commented on Dockerfile.test-data-generator:
```diff
@@ -0,0 +1,29 @@
+    # In the test setup (ref: docker-compose.test.yml), the 'sut' service which is mocking the gpg-s3sync functionality
+    # is run as root. This can be changed if you extend the base image or by adding USER directive to the Dockerfile
+    mkdir -p "$DATA_PATH" && \
+    chown daemon:daemon "$DATA_PATH"; \
```
Mix up of `;` and `&&` - I'd have thought we wouldn't want to proceed if the `chown` fails?  Or might we expect a non-zero status code in certain scenarios?

abulford commented on docker-compose.test.yml:
```diff
@@ -1,30 +1,49 @@
     volumes:
       - data:/data/plain
-      - encrypted-data:/data/encrypted
+      - ./run-test.sh:/run-test.sh
```
I think this script will already exist in the Docker image at `/app/run-test.sh`?

codemedic replied:
No .. it is ignored in `.dockerignore` Line 4, for the sake of the main `Dockerfile`

abulford replied:
Ah, I see.  This would be solved by putting the test-related stuff into a separate directory.  I nearly suggested that earlier for the sake of tidiness, but thought it would be tricky to do if the test scripts are using some files in common with the main Docker image.  Maybe there's some kind of re-arranging we could do in the future.

abulford commented on entrypoint.sh:
```diff
@@ -82,21 +98,24 @@ restore() {
 
 list-backups() {
+    # source-path must exist
+    [ -e "${SOURCE_PATH}" ] ||
```
Should this (and line 110) check for the path being a readable/executable directory?

abulford commented on run-test.sh:
```diff
@@ -11,55 +11,110 @@ set_debug
+            # if s3sync is enabled, then remove local encrypted files so that they are pulled down from S3, during
+            # restore attempts below
+            if s3sync is_enabled; then
+                rm -f "${ENCRYPTED_PATH}/"*.txt.gpg
```
Doesn't the clean up on line 57 make this unnecessary?

codemedic replied:
The one on line 57 is for cleaning up any residual data from a failed previous run; relevant only while developing/testing locally.
The latter / this one is for testing the specific functionality of download from S3 when the file is missing in `ENCRYPTED_PATH`

abulford commented on run-test.sh:
```diff
@@ -11,55 +11,110 @@ set_debug
+        # Using one file from the above set of tests, which has already been restored into SOURCE_PATH, check if the
+        # restore will fail if the destination file already exists. If the command succeeds, i.e the file was
+        # overwritten, then it is treated as a failure.
+        if /app/entrypoint.sh restore "data-01.txt.gpg"; then
```
Potentially this could be done before the clean up (line 84)?  Which is maybe a little more representative, as the file will have been created by the application rather than the syncer.

abulford commented on run-test.sh:
```diff
@@ -11,55 +11,110 @@ set_debug
-run
+cmd="$1"; shift
+case "$cmd" in
+    wait)
```
I know this was used during development, but is there still a need for it?  If so a comment would be useful.

abulford reviewed (commented):
Looks good, few minor comments, feel free to incorporate them if you agree or just merge if you disagree/can't be bothered.

codemedic committed:
Fixing `&&` & `;` mixup

codemedic committed:
Extra checks for SOURCE_PATH & ENCRYPTED_PATH

codemedic committed:
More tweaks based on PR comments

abulford commented on run-test.sh:
```diff
@@ -80,6 +80,14 @@ run_test() {
             local filepath="${SOURCE_PATH}/data-$(printf '%02d' "$i").txt"
             local filename=$(basename "$filepath")
 
+            # Using one file from the above set of tests, which has already been restored into SOURCE_PATH, check if
```
This comment's doesn't make much sense any more.

codemedic committed:
More info in permission check error messages

Ownership information was lost in an earlier change; added that back.
Also updated some comments.
